pydocstringformatter/run.py

```diff
@@ -113,11 +100,129 @@ def format_file(self, filename: Path) -> bool:
 
         return is_changed
 
-    def format_files(self, filepaths: list[Path]) -> bool:
-        """Format a list of files."""
+    def format_file_tokens(
+        self, tokens: list[tokenize.TokenInfo], filename: Path
+    ) -> tuple[list[tokenize.TokenInfo], bool]:
+        """Format a list of tokens.
+
+        tokens: List of tokens to format.
+        filename: Name of the file the tokens are from.
+
+        Returns:
+            A tuple containing [1] the formatted tokens in a list
+            and a boolean indicating if the tokens were changed.
+
+        Raises:
+            RuntimeError:
+                If the formatters are not able to get to a stable
+                result after 3 iterations.
+
+                It reports what formatters are still modifying the tokens.
+        """
+        formatted_tokens: list[tokenize.TokenInfo] = []
         is_changed = False
 
-        for file in filepaths:
-            is_changed = self.format_file(file) or is_changed
+        for index, tokeninfo in enumerate(tokens):
+            new_tokeninfo = tokeninfo
 
-        return is_changed
+            if _utils.is_docstring(new_tokeninfo, tokens[index - 1]):
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo
+                )
+                is_changed = is_changed or bool(changers)
+
+                # Try again to get a stable result, skipping the formatters that
+                # didn't change the tokens in former passes
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo,
+                )
+
+                if not stable:
+                    if len(changers) > 1:
+                        msg = "Conflicting formatters: " + ", ".join(changers)
+                    else:
+                        msg = "Formatter: " + changers[0]
+                        msg += "Is not able to make stable changes on:\n"
+                        msg += new_tokeninfo.string + "\n"
+
+                    if len(changers) == 2:
+                        diff = compare_formatters(
+                            new_tokeninfo,
+                            self.enabled_formatters[changers[0]],
+                            self.enabled_formatters[changers[1]],
+                            str(filename),
+                        )
+
+                        msg += f"\n{diff}"
+
+                    raise RuntimeError(msg)
+
+            formatted_tokens.append(new_tokeninfo)
+
+        return formatted_tokens, is_changed
+
+    def get_enabled_formatters(self) -> dict[str, _formatting.Formatter]:
```

Comment:
It still applies the style by filtering out the formatters whose style "is not active" (in the configuration)

Reply:
Hm, but we were doing that already though right? This seems to add some methods and attribute for something that we currently do in a for loop of 3/4 lines

Reply:
Well to be more explicit, it replaces L72-L79 in the parent (loop inside 'format file'). For this method.

In my opinion it abstracts away the "getting the enabled formatters, provided the current configuration" and makes sure it gets called only once during initialization (instead of for every file).

So i think It trades off some "lines" for conceptual clarity on the logic of the formatter.
If this is something you are not interested in, we could move it back to the main function. LMK if you would like me to do that

Reply:
You're right! Doing this only once makes sense.

That said, perhaps we should extract it into its own PR? It should be much easier to review and get it in.

Do you know why the tests are failing?

Reply:
1. Wohooo
2. Moved it to PR: https://github.com/DanielNoord/pydocstringformatter/pull/178
3. I am not 100% sure, but I think the `type(CLASSS)` synmtax was added in 3.9, pre 3.9 I THINK `typing.Type[CLASS]` should work (not sure why the linters did not pick that up) ...

Reply:
Yup that was it